---
title: "Differentially Expressed Gene"
author: "Narad Mishra"
date: "November 28, 2017"
output: html_document
---

```{r, warning=FALSE, message=FALSE}
source("https://bioconductor.org/biocLite.R")
#biocLite("ALL")
library(ALL)
#biocLite("limma")
library(limma)
#biocLite("Biobase")
library(Biobase)
library(DT)
data(ALL)
pd<-pData(ALL)
#limma disallows special characters i nfactor names

#convenient wrapper to make names compatible with R internals
mol.biol<-make.names(pd$mol.biol)
groups<-factor(mol.biol, levels = unique(mol.biol))
design<-model.matrix(~0 + groups)   #multifactorial design matrix
colnames(design)<-levels(groups)
datatable(design)

#defining what to compare
contrast<-makeContrasts(BCR.ABLvsNEG = BCR.ABL - NEG, levels = design)


#normalizing gene expression before differential expression
M<-normalizeQuantiles(exprs(ALL))
fit<-lmFit(M, design)
fit.c<-contrasts.fit(fit, contrast)
fit.c<-eBayes(fit.c)

#all results of differential expression analysis
tt<-topTable(fit.c, number=Inf, coef = "BCR.ABLvsNEG", adjust.method="none", p.value = 0.05)

```

1. Write a function that takes a vector of signiffcantly differentially expressed probes and annotate them by chromosomes they are located on. To do this, you mightfind the "get" function useful. For example: get(paste("hgu95av2", "CHRLOC", sep="")) returns the CHRLOC map for the hgu95av2 annotation package, assuming it has been loaded already by library().

```{r, message=FALSE, warning=FALSE}
probe<-data.frame(cbind(row.names(tt), tt$P.Value))
colnames(probe)<-c("probe_id",  "pvalue")

source("https://bioconductor.org/biocLite.R")
#biocLite("hgu95av2.db")
library(hgu95av2.db)
library(dplyr)
ALL@annotation # Check which microarray platform was used. Or, annotattion(ALL)

xx <- as.list(hgu95av2SYMBOL) # Get full mapping table for probes to gene names
gene <- unlist(xx[row.names(tt)])
probe$gene<-gene

sys<-function(probe.name){
  chromosome<-suppressWarnings(as.character(toTable(hgu95av2CHR[probe.name])[2]))
  if(chromosome=="character(0)") chromosome=NA
  return(chromosome)
}

probes<-rownames(tt)
sigdiff<-tt
sigdiff$Chromosome<-sapply(probes,sys)
head(sigdiff)

```



2. Use Fisher's Exact test to test whether any chromosomes are significantly enriched or depleted for differentially expressed genes.

```{r, message=FALSE, warning=FALSE}
fisher<-topTable(fit.c, number=Inf, coef = "BCR.ABLvsNEG", adjust.method="none")
probe<-rownames(fisher)
fisher2<-fisher
fisher2$chromosome<-sapply(probe,sys)

test<-function(chromosome,n=nrow(fit.c),data=fisher2){
sig.probes<-(data$P.Value<0.05)*1
inchro<-(data$chromosome==chromosome)*1
tbl<-t(table(inchro,sig.probes))
test<-fisher.test(tbl)
return(list(p.val=test$p.value,OR=test$estimate))
}
chromosome<-c(1:22, "X", "Y")
fisher.result<-as.data.frame(t(sapply(chromosome,test)))
df<-data.frame(chromosome=chromosome,p.val=as.numeric(fisher.result$p.val),OR=as.numeric(fisher.result$OR))

# odds ratio above 1 is enriched and below 1 is depleted 
a<-df[df$p.val<0.05,]

enriched<-which(a$OR>1)
depletes<-which(a$OR<1)

# The list of chromosome that are enrichmed based on odds ratio greater than one.
enriched
depletes


```


3. Correct for Fisher's exact p-values for multiple testing using all methods you know.

```{r warning=FALSE, message=FALSE}
library(stats)
df$bonferroni<-p.adjust(df$p.val, method="bonferroni")

df$holm<-p.adjust(df$p.val, method="holm")

df$hochberg<-p.adjust(df$p.val, method="hochberg")

df$fdr<-p.adjust(df$p.val, method="fdr") 

df$BH<-p.adjust(df$p.val, method="BH")



```



4. Repeat the same calculations using hypergeometric test, binomial test, permutation test. You may find this presentation helpful: http://jura.wi.mit.edu/bio/education/hot_topics/enrichment/Gene_list_ enrichment_Mar10.pdf. Do not adjust for multiple testing.

```{r}
library(perm)
difftests<-function(chromosome,n=nrow(fit.c),data=fisher2,method){
  sig.probes<-(data$P.Value<0.05)*1
  inchro<-(data$chromosome==chromosome)*1
tbl<-t(table(inchro,sig.probes))
  mat<-matrix(c(tbl[4],tbl[3],tbl[2],tbl[1]),nrow=2)
  
  test<-switch(method,
         "hypergeometric"= min(1 - cumsum(dhyper(0:(tbl[4]-1), mat[3], mat[4], colSums(mat)[1] ) )),
         "binomial"= binom.test(mat[1],n=colSums(mat)[1],p= mat[3]/(colSums(mat)[2]))$p.value,
         "permutation"= permTS(data$P.Value~inchro,alternative="two.sided",exact=TRUE)$p.value)
  
  return(test)
}

chromosome<-c(1:22, "X", "Y")
df1<-df
df1$hyper=sapply(chromosome,function(x) difftests(x,method="hypergeometric"))
df1$binomial=sapply(chromosome,function(x) difftests(x,method="binomial"))
df1$permutation=sapply(chromosome,function(x) difftests(x,method="permutation"))

```

5. Summarize the results in one table. Rows are chromosomes, columns are p-values. Column names should similar to "Fisher's exact p-value", "Fisher's exact Bonferroni adjusted", "Fisher's exact FDR adjusted", "...", "hypergeometric p-value", "binomial p-value", "permutation p-value". 

```{r}
#data df from fisher test
df<-df1[,c("chromosome","p.val", "bonferroni", "holm", "hochberg", "fdr", "BH",  "hyper", "binomial", "permutation")]
colnames(df)<-c("chromosome", "Fisher.p.value", "Fisher.Bonferroni.adjust",
                "fisher.holm.adjust","fisher.hochberg.adjust","fisher.fdr.adjust",
                "fisher.BH.adjusted", "binom.pvalue","hyper.pvalue","permutation.pval")

head(df)
```


6. Interpret the differences in raw p-values from different tests. 

* The p-values from the binomial and the fisher's exact were pretty similar. However, the p-values for the few chromosomes for the hypergeometric were sparkly different compared to the other tests. The hypergeometric test (according to the slides) is shown to only look for the enriched ones whereas the other tests look for both enriched and depleted.

7. Interpret the differences among different methods for adjustment for multiple testing.

* Looking at the different adjusted p-values, BH-adjusted and fdr-adjusted are exactly the same and I am much surprised by seeing the p-values for adjusted holm, bonferroni, and hochberg, that almost all the p-values are 1. They seems to be much more conservatives compared to fdr-adjusted and other pvalue obtained from different tests. 